kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java

```diff
@@ -288,6 +309,30 @@ private boolean checkRotationOrAppend(
     return true;
   }
 
+  /**
+   * Use a temporary writer to check the record for DataExceptions ahead of writing.
```

Comment:
do we need to check ahead? seems expensive

Reply:
`DataException` is thrown by [fromConnectData()](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/349/files#diff-16f592a34da31ec963db3a13a0726eeb8b9d6176475ebe05ba84d1c760aa3ff4R71-R72 ), so I don't think we need this check ahead. just try to [write here](https://github.com/confluentinc/kafka-connect-storage-cloud/pull/349/files#diff-4e620d031f99e2962510e761f99cae963938454ad0e71a6a79d1b1cf3c7ffc99L501) and catch the data exception

Reply:
I can't open you diff links for some reason, let's try file links. Overall I agree with that we shouldn't have a check ahead and I tried not to have it. I believe you mention the existing position of the write, the reason I wasn't able to handle this there is because the logic before that write call affects global vars and we'd have to manually restore those states after the catch, which seemed very complex. This seemed the most readable solution, lmk if you have ideas.

Reply:
just write the record first, and if it succeeds then change all of the global variables. should be pretty simple

Reply:
The current `TopicPartitionWriter` implementation causes issues with that approach:
1. The writer used to write the record is itself retrieved based on the configuration of the global vars, so this doesn't allow us to write the record ahead of setting the global vars. 

    The writer ahead of the write is retrieved on [L#501](https://github.com/dosvath/kafka-connect-storage-cloud/blob/4cd601138a838f61fba6dbd4f314a470c01c7dac/kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java#L501). If you trace through the method calls made by that retrieval you will see one of the methods used `getCommitFilename` uses the global var  `startOffsets`.

2. Additional global variables are set at a higher level ahead of the [writeRecord](https://github.com/dosvath/kafka-connect-storage-cloud/blob/4cd601138a838f61fba6dbd4f314a470c01c7dac/kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java#L275) call. The write call is dependent on these to be set ahead. [currentEncodedPartition here](https://github.com/dosvath/kafka-connect-storage-cloud/blob/4cd601138a838f61fba6dbd4f314a470c01c7dac/kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java#L269), [currentSchemas here](https://github.com/dosvath/kafka-connect-storage-cloud/blob/4cd601138a838f61fba6dbd4f314a470c01c7dac/kafka-connect-s3/src/main/java/io/confluent/connect/s3/TopicPartitionWriter.java#L216). 
Setting these ahead of the write is currently required for the `TopicPartitionWriter` to function properly. 

Refactoring the `TopicPartitionWriter` class to behave differently can be done but would first require a major refactoring PR without any DLQ changes, and would need to be thoroughly tested to eliminate possible regressions.

Reply:
If it's any consolation, keep in mind the above temporary writer does not write immediately to a file, only if commit is called, and if it's a correct record the result will be cleaned up by JVM when the tmp writer is closed.

Reply:
I will look more into this, but does writing a bad record, pollute the writer? any chance we could just try writing, fail, and keep moving?

Reply:
I believe the writer itself wouldn't be the issue in that case, it could be used for further writing, but the record count in the file would not align with the offsets in the TPW class and cause misalignment in the globals mentioned, resulting in failing unit tests/ altered behavior of the class.

Reply:
Unfortunately we can't introduce a performance regression per record if anyone enables error reporting. This performance tax is not an option. That's the bad news. 

The good news is that error inspection happens per record. So we can think like a compiler, meaning we can do register allocation (move variables around in local copies if we have to) and instruction reordering, keeping the instructions that need to happen before `writer.write(record)` before (in which case we'll track and revert the state they are changing if an error happens), while leaving any other instructions that mutate the state of the TPW after, to be executed only if a write succeeds.